app/services/platforms/twitch/tags.ts

```diff
@@ -0,0 +1,157 @@
+ * we can request subsequent pages of tags.
+ */
+interface PaginatedResponse {
+  items: Array<TwitchTag>;
+  cursor: string;
+  /*
+   * We need to track oldCursor since apparently Twitch API doesn't
+   * unset `cursor` in the last response. We check for equality later.
+   */
+  oldCursor: string;
+}
+
+const requestTags = (
+  headers: TwitchRequestHeaders,
+  cursor: string,
+): Observable<PaginatedResponse> =>
+  ajax
+    .getJSON<TwitchTagsResponse>(
+      `https://api.twitch.tv/helix/tags/streams?first=100&after=${cursor}`,
+      headers,
+    )
+    .pipe(
+      map(response => ({
+        cursor: response.pagination.cursor,
+        items: response.data,
+        oldCursor: cursor,
+      })),
+    );
+
+/**
+ * Fetch all available tags that Twitch provides that are not
+ * automatically generated. This will use the provided pagination
+ * to request the whole dataset of tags.
+ *
+ * @param headers Headers including OAuth Token and App ID
+ */
+// FIXME: Twitch Pagination seems broken regarding cursor
+export const getAllTags = (headers: TwitchRequestHeaders): Promise<Array<TwitchTag>> =>
+  requestTags(headers, '')
+    .pipe(
+      // expand(
+      //   ({ cursor, oldCursor }) =>
+      //     cursor && cursor !== oldCursor ? requestTags(headers, cursor) : empty(),
+      //   1,
+      // ),
+      map(({ items }) => items.filter(tag => !tag.is_auto)),
+    )
+    .toPromise();
+
+const getLabelFor = (tag: TwitchTag, locale: string): string =>
+  tag.localization_names[locale.toLowerCase()] || tag.localization_names['en-us'];
+
+const assignLabels = (locale: string) => (tags: Array<TwitchTag>): Array<TwitchTagWithLabel> =>
+  tags.map(tag => ({
+    ...tag,
+    name: getLabelFor(tag, locale),
+  }));
+
+const byName = contramap((tag: TwitchTagWithLabel) => tag.name, ordString);
+const sortByName = sortBy1(byName, []);
+
+export const prepareOptions = (
+  locale: string,
+  tags: Array<TwitchTag> | undefined,
+): Array<TwitchTagWithLabel> =>
+  fromNullable(tags)
+    .map(
+      compose(
+        sortByName,
+        assignLabels(locale),
+      ),
+    )
+    .getOrElse([]);
+
+/**
+ * Add a parameter, either "add" or "remove" with tag IDs separated by commas
+ */
+const addParam = (op: 'add' | 'remove', tags: Array<TwitchTagWithLabel>) => {
+  if (!tags.length) {
+    return {};
+  }
+
+  return { [op]: tags.map(tag => tag.tag_id).join(',') };
+};
+
+export const updateTags = (headers: TwitchRequestHeaders) => (tags: Array<TwitchTagWithLabel>) => (
+  streamId: string,
+) => {
+  const toAdd = tags;
+
+  // TODO: how to track removals if we can't even get the active list w/o going live
+  const toRemove: Array<TwitchTagWithLabel> = [];
+
+  const params = {
+    ...addParam('add', toAdd),
+    ...addParam('remove', toRemove),
+  };
+
+  return ajax
+    .put(
+      `https://api.twitch.tv/helix/tags/streams?broadcaster_id=${streamId}`,
+      JSON.stringify(params),
+      headers,
+    )
+    .toPromise();
+};
```

Comment:
we traditionally use the `fetch` api to handle our HTTP requests, and although i'm a personal fan of ajax it seems like an antipattern here

Reply:
The rest of this file uses Rx, and we need that since we're going to be doing recursive requests. This particular case is for consistency. We also have proper error handling (as opposed to `fetch`) when a status code is error or JSON parsing fails, for free.
We might also move the delay here if it turns out there's no way we can invoke `updateTags` reliably without one, we'll need Rx here too. Hopefully makes sense?